### PR TITLE
refactor(linux): Use `km_core_actions` struct instead of queue 🌱

### PR DESCRIPTION
**Pull request by ermshiperete**
Implements #10353.

# User Testing

## Preparations

- The tests can be be run on any Linux platforms (running with X11). Note that Firefox and Chromium have to be installed from .deb, the snap version won't work because it's missing the ibus patch.

- Install Anki
  - Download [Qt5 version of Anki](https://apps.ankiweb.net/#linux)
  - Run the following commands
    ```bash
    sudo apt update
    sudo apt install zstd
    tar xaf Downloads/anki-2.*-linux-qt5.tar.zst
    cd anki-2.*-linux-qt5
    sudo ./install.sh
    sudo apt install libxcb-xinerama0
    ```
  - When testing in Anki, select e.g. "Tools"/"Study Deck"

- Install patched ibus by running these commands:
  ```
  sudo add-apt-repository ppa:keymanapp/keyman-alpha
  sudo apt upgrade
  ```
  - afterwards running `dpkg -l ibus` should show a version number with `sil` in the version string, e.g. `1.5.28-3sil2~jammy`
 
- [Install build artifacts of this PR](https://github.com/keymanapp/keyman/wiki/How-to-test-artifacts-from-pull-requests-for-Keyman-for-Linux)

- Reboot

- Install the following keyboards in Keyman:
  * [IPA (SIL)](https://keyman.com/keyboards/sil_ipa)
  * [Korean KORDA Jamo (SIL)](https://keyman.com/keyboards/sil_korda_jamo)
  * [Khmer Angkor](https://keyman.com/keyboards/khmer_angkor)
  * [Vedic Sanskrit Devanagari Phonetic (ITRANS)](https://keyman.com/keyboards/itrans_devanagari_sanskrit_vedic)

## SUITE_WRITER: LibreOffice Writer

Open LibreOffice Writer.

### Tests

- **TEST_WRITER_IPA:** Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".
- **TEST_WRITER_KO:** Switch to "Korean KORDA Jamo (SIL)" keyboard. Type `han<space>geul<space>`. Verify that the result is "한글".
- **TEST_WRITER_HI:** Switch to "Vedic Sanskrit Devanagari Phonetic (ITRANS)" keyboard. Type `shrI`. Verify that the result is "श्री". (If the result looks wrong, select all text and change the font to "Siddhanta")
- **TEST_WRITER_KM:** Switch to "Khmer Angkor" keyboard. Select "Khmer Mondulkiri" font. Type `xEjmr`. Verify that the output is "ខ្មែរ".
- **TEST_WRITER_KM_BS:** With the output "ខ្មែរ" still being displayed, press Backspace key 5 times. Verify that after each keystroke a part of the string is removed and after the 5th backspace the entire word is deleted.

## SUITE_GEDIT: gedit

Open gedit.

### Tests

- **TEST_GEDIT_IPA:** Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".
- **TEST_GEDIT_KO:** Switch to "Korean KORDA Jamo (SIL)" keyboard. Type `han<space>geul<space>`. Verify that the result is "한글".
- **TEST_GEDIT_HI:** Switch to "Vedic Sanskrit Devanagari Phonetic (ITRANS)" keyboard. Type `shrI`. Verify that the result is "श्री".
- **TEST_GEDIT_KM:** Switch to "Khmer Angkor" keyboard. Type `xEjmr`. Verify that the output is "ខ្មែរ".
- **TEST_GEDIT_KM_BS:** With the output "ខ្មែរ" still being displayed, press Backspace key 5 times. Verify that after each keystroke a part of the string is removed and after the 5th backspace the entire word is deleted.

## SUITE_FIREFOX: Firefox

Open https://keyman.com/keyboards in Firefox.

**NOTE:** If the output looks wrong, copy the text from the browser and paste it into gedit and verify that it there.

### Tests

- **TEST_FIREFOX_IPA:** Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".
- **TEST_FIREFOX_KO:** Switch to "Korean KORDA Jamo (SIL)" keyboard. Type `han<space>geul<space>`. Verify that the result is "한글".
- **TEST_FIREFOX_HI:** Switch to "Vedic Sanskrit Devanagari Phonetic (ITRANS)" keyboard. Type `shrI`. Verify that the result is "श्री".
- **TEST_FIREFOX_KM:** Switch to "Khmer Angkor" keyboard. Type `xEjmr`. Verify that the output is "ខ្មែរ".

**NOTE:** Backspace with Firefox won't work correctly because of #9971).

## SUITE_CHROMIUM: Chrome/Chromium

Open https://keyman.com/keyboards in Chrome/Chromium browser.

**NOTE:** If the output looks wrong, copy the text from the browser and paste it into gedit and verify that it there.

### Tests

- **TEST_CHROMIUM_IPA:** Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".
- **TEST_CHROMIUM_KO:** Switch to "Korean KORDA Jamo (SIL)" keyboard. Type `han<space>geul<space>`. Verify that the result is "한글".
- **TEST_CHROMIUM_HI:** Switch to "Vedic Sanskrit Devanagari Phonetic (ITRANS)" keyboard. Type `shrI`. Verify that the result is "श्री".
- **TEST_CHROMIUM_KM:** Switch to "Khmer Angkor" keyboard. Type `xEjmr`. Verify that the output is "ខ្មែរ".
- **TEST_CHROMIUM_KM_BS:** With the output "ខ្មែរ" still being displayed, press Backspace key 5 times. Verify that after each keystroke a part of the string is removed and after the 5th backspace the entire word is deleted.

## SUITE_TERMINAL: gnome-terminal

Open Terminal.

### Tests

- **TEST_TERMINAL_IPA:** Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".

**NOTE:** the other tests for gnome-terminal won't work because of #10481 

## SUITE_ANKI: Anki

Open Anki.

### Tests

- **TEST_ANKI_IPA:** Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".
- **TEST_ANKI_KO:** Switch to "Korean KORDA Jamo (SIL)" keyboard. Type `han<space>geul<space>`. Verify that the result is "한글".
- **TEST_ANKI_HI:** Switch to "Vedic Sanskrit Devanagari Phonetic (ITRANS)" keyboard. Type `shrI`. Verify that the result is "श्री" (If the result looks wrong, copy/paste it in text editor and verify there).
- **TEST_ANKI_KM:** Switch to "Khmer Angkor" keyboard. Type `xEjmr`. Verify that the output is "ខ្មែរ".
- **TEST_ANKI_KM_BS:** With the output "ខ្មែរ" still being displayed, press Backspace key 5 times. Verify that after each keystroke a part of the string is removed and after the 5th backspace the entire word is deleted.

## SUITE_SEARCHBAR: Searchbar in gnome-shell

Press Windows key to open searchbar.

### Tests

- **TEST_SEARCHBAR_IPA:** Switch to "IPA (SIL)" keyboard. Type `n>`. Verify that the result is "ŋ".
- **TEST_SEARCHBAR_KO:** Switch to "Korean KORDA Jamo (SIL)" keyboard. Type `han<space>geul<space>`. Verify that the result is "한글".
- **TEST_SEARCHBAR_HI:** Switch to "Vedic Sanskrit Devanagari Phonetic (ITRANS)" keyboard. Type `shrI`. Verify that the result is "श्री".
- **TEST_SEARCHBAR_KM:** Switch to "Khmer Angkor" keyboard. Type `xEjmr`. Verify that the output is "ខ្មែរ".
- **TEST_SEARCHBAR_KM_BS:** With the output "ខ្មែរ" still being displayed, press Backspace key 5 times. Verify that after each keystroke a part of the string is removed and after the 5th backspace the entire word is deleted.

### CODE REVIEW

File: linux/ibus-keyman/src/engine.c [hunk 1]
```diff
@@ -745,25 +701,18 @@ process_end_action(IBusKeymanEngine *keyman) {
...
-      // We have an old ibus version without prefilter support, or a client that does support
-      // surrounding text. In either case we return FALSE because we emitted a keystroke
-      // so that the processing of the event will continue.
-      return FALSE;
-    }
+    // compliant app, or unpatched ibus
+    // If we have an old ibus version without prefilter support, or a client that does support
+    // surrounding text we already did everything that needs to be done including forwarding
+    // a keystroke. So we can return TRUE to stop processing of this event.
+      return TRUE;
```
Comment (darcywong00) [inline linux/ibus-keyman/src/engine.c:715]: Does this mean finish_process_actions() never returns FALSE now?

File: linux/ibus-keyman/src/engine.c [hunk 2]
```diff
@@ -558,160 +560,113 @@ static void commit_string(IBusKeymanEngine *keyman, const gchar *string)
...
-  km_core_option_item *keyboard_opts = g_new0(km_core_option_item, 2);
-  memmove(&(keyboard_opts[0]), action_item->option, sizeof(km_core_option_item));
-  km_core_status event_status = km_core_state_options_update(keyman->state, keyboard_opts);
-  if (event_status != KM_CORE_STATUS_OK) {
-    g_warning("%s: problem saving option for km_core_keyboard", __FUNCTION__);
+static void
+process_persist_action(IBusEngine *engine, km_core_option_item *persist_options) {
+  if (persist_options == NULL) {
+    return;
   }
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/engine.c:624]: `persist_options` should never be null. If empty, it should be an array with a single item, the `is_core_options_end()` item. So this should be a `g_assert`

```suggestion
  g_assert(persist_options != NULL);
```

File: linux/ibus-keyman/src/engine.c [hunk 3]
```diff
@@ -558,160 +560,113 @@ static void commit_string(IBusKeymanEngine *keyman, const gchar *string)
...
-  // Put the keyboard option into DConf
-  if (action_item->option->key != NULL && action_item->option->value != NULL) {
-    g_message("%s: Saving keyboard option to DConf", __FUNCTION__);
-    // Load the current keyboard options from DConf
-    keyman_put_options_todconf(
-        keyman->kb_name, keyman->kb_name, (gchar *)action_item->option->key, (gchar *)action_item->option->value);
+  IBusKeymanEngine *keyman = (IBusKeymanEngine *)engine;
+  for (km_core_option_item *option = persist_options; !is_core_options_end(option); option++) {
+    // Put the keyboard option into DConf
+    if (option->key != NULL && option->value != NULL) {
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/engine.c:629]: This should probably be a g_assert as well; we should never be passing invalid data to you from Core.

File: linux/ibus-keyman/src/engine.c (hunk 1 above)
Comment (mcdurdin) [reply linux/ibus-keyman/src/engine.c:715]:
```suggestion
    return TRUE;
```

**Review by mcdurdin — APPROVED** [review]
LGTM. Cleaner. No marker shenanigans!

**Comment by bharanidharanj** [comment]
## SUITE_WRITER: LibreOffice Writer

* **TEST_WRITER_IPA (PASSED):** Tested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2125.1+mantic1) on Ubuntu 23.10 Mantic Minotaur OS (vm) and here is my observation: 1. Installed the above mentioned languages (IPA (SIL), Korean KORDA Jamo, Khmer Angkor and Vedic Sanskrit Devanagari..) using Keyman Configuration dialog. 2. Opened LibreOffice Writer. 3. Verified that the result ŋ appeared on the document. 

   

* **TEST_WRITER_KO (PASSED):** 1. Verified that the result appeared on the document. 
* **TEST_WRITER_HI (PASSED):** 1. Verified that the result श्री appeared on the document. 
* **TEST_WRITER_KM (FAILED):** 1. Noticed that it is showing the wrong output. 
* **TEST_WRITER_KM_BS (FAILED):** 1. Did not pass the test since the previous one failed.

**Comment by bharanidharanj** [comment]
# Test Results

## SUITE_GEDIT: gedit

* **TEST_GEDIT_IPA (OPEN):** Tested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2125.1+mantic1) on Ubuntu 23.10 Mantic Minotaur OS (vm) (Ubuntu on Xorg - x11)  and here is my observation: 1. Installed the above-mentioned languages (IPA (SIL), Korean KORDA Jamo, Khmer Angkor, and Vedic Sanskrit Devanagari..) using Keyman Configuration dialog. 2. Opened Text Editor. 3. Noticed that it is showing the wrong output.

   


* **TEST_GEDIT_KO (OPEN):** 1. Verified that it is showing the correct output on the document. 
* **TEST_GEDIT_HI (OPEN):** 1. Noticed that it is showing the wrong output on the document.
* **TEST_GEDIT_KM (OPEN):** 1. Noticed that it is showing the wrong output. 
* **TEST_GEDIT_KM_BS (OPEN):** 1. Did not pass the test since the previous one is failed.

File: linux/ibus-keyman/src/engine.c (hunk 1 above)
Comment (ermshiperete) [reply linux/ibus-keyman/src/engine.c:715]: > Does this mean finish_process_actions() never returns FALSE now?

yes. And this means the method doesn't have to return a bool. Changed. Thanks!

**Comment by bharanidharanj** [comment]
# Test Results

## SUITE_FIREFOX: Firefox

* **TEST_FIREFOX_IPA (PASSED):** Tested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2125.1+mantic1) on Ubuntu 23.10 Mantic Minotaur OS (vm) (Ubuntu on Xorg - x11)  and here is my observation: 1. Installed the above-mentioned languages (IPA (SIL), Korean KORDA Jamo, Khmer Angkor and Vedic Sanskrit Devanagari..) using Keyman Configuration dialog. 2. Opened Mozilla Firefox browser. 3. Opened https://keyman.com/keyboards link. 4. Switched to sil_ipa keyboard. 5. Verified it shows the expected result.

   

* **TEST_FIREFOX_KO (PASSED):** 1. Verified that it is showing the correct output on the document. 
   


* **TEST_FIREFOX_HI (PASSED):** 1. Verified that it is showing the correct output on the document. 

   

* **TEST_FIREFOX_KM (PASSED):** 1. Verified that it is showing the correct output on the document.

**Comment by bharanidharanj** [comment]
# Test Results

## SUITE_CHROMIUM: Chromium

* **TEST_CHROMIUM_IPA (FAILED):** Tested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2125.1+mantic1) on Ubuntu 23.10 Mantic Minotaur OS (vm) (Ubuntu on Xorg - x11)  and here is my observation: 1. Installed the above mentioned languages (IPA (SIL), Korean KORDA Jamo, Khmer Angkor and Vedic Sanskrit Devanagari..) using Keyman Configuration dialog. 2. Opened Chromium browser. 3. Opened https://keyman.com/keyboards link. 4. Switched to sil_ipa keyboard. 5. Noticed that it is showing the wrong output. FYI, I have installed Mozilla Firefox and Chrome browsers from .deb. 


   

* **TEST_CHROMIUM_KO (FAILED):** 1. Noticed that it is showing the wrong output.

   

* **TEST_CHROMIUM_HI (FAILED):** 1. Noticed that it is showing the wrong output.

   

* **TEST_CHROMIUM_KM (FAILED):** 1. Noticed that it is showing the wrong output.

   

* **TEST_CHROMIUM_KM_BS (FAILED):** 1. Did not pass the test since the previous one is failed.

**Comment by ermshiperete** [comment]
@bharanidharanj Do you have the patched ibus installed (`dpkg -l ibus` should show a version with `sil` in the version string). If not you'll have to add the keyman-alpha ppa (`sudo add-apt-repository ppa:keymanapp/keyman-alpha`) and then upgrade the  packages (`sudo apt upgrade`). Afterwards you'll have to install Keyman for this PR again because the ppa contains a Keyman version with a higher version number, so it got upgraded.

**Comment by bharanidharanj** [comment]
## SUITE_SEARCHBAR: Searchbar in gnome-shell

* **TEST_SEARCHBAR_IPA (FAILED):** Tested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2125.1+mantic1) on Ubuntu 23.10 Mantic Minotaur OS (vm) (Ubuntu on Xorg - x11)  and here is my observation: 1. Installed the above mentioned languages (IPA (SIL), Korean KORDA Jamo, Khmer Angkor and Vedic Sanskrit Devanagari..) using Keyman Configuration dialog. 2. Pressed Windows key to open search bar. 3. Switched to sil_ipa keyboard. 5. Observed that there was no output displayed in the search bar. 

   


* **TEST_SEARCHBAR_KO (PASSED):** Noticed that it is showing the correct output.

   

* **TEST_SEARCHBAR_HI (FAILED):** Observed that there was no output displayed in the search bar. 

   


* **TEST_SEARCHBAR_KM (PASSED):** Noticed that it is showing the correct output.

   


* **TEST_SEARCHBAR_KM_BS (PASSED):** 1. Pressing Backspace key 5 times, removed the entire khmer_angkor word.

**Comment by bharanidharanj** [comment]
# Test Results

## SUITE_ANKI: Anki

* **TEST_ANKI_IPA (PASSED):** Tested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2125.1+mantic1) on Ubuntu 23.10 Mantic Minotaur OS (vm) (Ubuntu on Xorg - x11)  and here is my observation: 1. Installed the above mentioned languages (IPA (SIL), Korean KORDA Jamo, Khmer Angkor and Vedic Sanskrit Devanagari..) using Keyman Configuration dialog. 2. Installed Anki application. 3. Opened it. Clicked Tools / Study Deck option. 4. In Study Deck window, Clicked the Search bar. 5. Switched to sil_ipa keyboard. 6. Typed 'n>' keys showed the expected output. 

   


* **TEST_ANKI_KO (PASSED):** Noticed that it is showing the correct output.

   

* **TEST_ANKI_HI (PASSED):** Noticed that it is showing the correct output.
   
   


* **TEST_ANKI_KM (PASSED):** Noticed that it is showing the correct output.

   

* **TEST_ANKI_KM_BS (PASSED):** 1. Pressing Backspace key 5 times, removed the entire khmer_angkor word.

**Comment by ermshiperete** [comment]
@bharanidharanj There seem to be problems with Ubuntu 23.10 which I will look into separately. Could you please re-test with Ubuntu 22.04 Jammy? Thanks!

Also, please make sure to test in **gedit** which is a separate app from Text Editor - you might have to install it explicitly with `sudo apt install gedit`.

@keymanapp-test-bot retest SUITE_WRITER  TEST_WRITER_KM TEST_WRITER_KM_BS SUITE_GEDIT  TEST_GEDIT_IPA TEST_GEDIT_HI TEST_GEDIT_KM TEST_GEDIT_KM_BS SUITE_CHROMIUM  TEST_CHROMIUM_IPA TEST_CHROMIUM_KO TEST_CHROMIUM_HI TEST_CHROMIUM_KM TEST_CHROMIUM_KM_BS SUITE_TERMINAL  TEST_TERMINAL_IPA SUITE_SEARCHBAR  TEST_SEARCHBAR_IPA TEST_SEARCHBAR_HI

**Comment by bharanidharanj** [comment]
> @bharanidharanj There seem to be problems with Ubuntu 23.10 which I will look into separately. Could you please re-test with Ubuntu 22.04 Jammy? Thanks!
> 
> Also, please make sure to test in **gedit** which is a separate app from Text Editor - you might have to install it explicitly with `sudo apt install gedit`.
> 
> @keymanapp-test-bot retest SUITE_WRITER TEST_WRITER_KM TEST_WRITER_KM_BS SUITE_GEDIT TEST_GEDIT_IPA TEST_GEDIT_HI TEST_GEDIT_KM TEST_GEDIT_KM_BS SUITE_CHROMIUM TEST_CHROMIUM_IPA TEST_CHROMIUM_KO TEST_CHROMIUM_HI TEST_CHROMIUM_KM TEST_CHROMIUM_KM_BS SUITE_TERMINAL TEST_TERMINAL_IPA SUITE_SEARCHBAR TEST_SEARCHBAR_IPA TEST_SEARCHBAR_HI

@ermshiperete Sure. I will do it.

**Comment by bharanidharanj** [comment]
## SUITE_CHROMIUM: Chrome/Chromium

* **TEST_CHROMIUM_IPA (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the sil_ipa Keyboard. 2. Opened Chromium browser. 3. Opened https://keyman.com/keyboards link. 4. Switched to sil_ipa keyboard. 5. Noticed that it is showing the correct output. 

* **TEST_CHROMIUM_KO (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the Korean_KORDA Keyboard. 2. Opened Chromium browser. 3. Opened https://keyman.com/keyboards link. 4. Switched to korean_korda keyboard. 5. Noticed that it is showing the correct output. 

* **TEST_CHROMIUM_HI (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the vedic_sanskrit Keyboard. 2. Opened Chromium browser. 3. Opened https://keyman.com/keyboards link. 4. Switched to vedic_sanskrit keyboard. 5. Noticed that it is showing the correct output. 


* **TEST_CHROMIUM_KM (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the khmer_angkor Keyboard. 2. Opened Chromium browser. 3. Opened https://keyman.com/keyboards link. 4. Switched to khmer_angkor keyboard. 5. Noticed that it is showing the correct output. 

* **TEST_CHROMIUM_KM_BS (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the sil_ipa Keyboard. 2. Opened Chromium browser. 3. Opened https://keyman.com/keyboards link. 4. Switched to sil_ipa keyboard. 5. With the output "ខ្មែរ" still being displayed, pressed the Backspace key 5 times. Verified that after each keystroke a part of the string is removed and after the 5th backspace the entire word was deleted.

**Comment by bharanidharanj** [comment]
## SUITE_TERMINAL: gnome-terminal

* **TEST_TERMINAL_IPA (FAILED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the sil_ipa keyboard. 2. Switched to sil_ipa keyboard. 3. Opened the terminal window. 4. Typed "n>" letters. 5. Noticed that it showed the wrong output.

**Comment by bharanidharanj** [comment]
## SUITE_SEARCHBAR: Searchbar in gnome-shell

* **TEST_SEARCHBAR_IPA (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the sil_ipa keyboard. 2. Switched to sil_ipa keyboard. 3. Opened the terminal window. 4. Typed "n>" letters. 5. Noticed that it showed the correct output.

   

* **TEST_SEARCHBAR_HI (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the vedic_sanskrit keyboard. 2. Switched to vedic_sanskrit keyboard. 3. Opened the terminal window. 4. Typed "shrI" letters. 5. Noticed that it showed the correct output.

**Comment by bharanidharanj** [comment]
## SUITE_GEDIT: gedit

* **TEST_GEDIT_IPA (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the sil_ipa keyboard. 2. Switched to sil_ipa keyboard. 3. Opened the terminal window. 4. Typed "n>" letters. 5. Noticed that it showed the correct output.


* **TEST_GEDIT_HI (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the vedic_sanskrit keyboard. 2. Switched to vedic_sanskrit keyboard. 3. Opened the terminal window. 4. Typed "n>" letters. 5. Noticed that it showed the correct output.


* **TEST_GEDIT_KM (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the khmer_angkor keyboard. 2. Switched to khmer_angkor keyboard. 3. Opened the terminal window. 4. Typed "n>" letters. 5. Noticed that it showed the correct output.


* **TEST_GEDIT_KM_BS (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the khmer_angkor Keyboard. 2. Opened Text Editor.  3. Switched to khmer_angkor keyboard. 4. With the output "ខ្មែរ" still being displayed, pressed the Backspace key 5 times. Verified that after each keystroke a part of the string is removed and after the 5th backspace the entire word was deleted.

**Comment by bharanidharanj** [comment]
# Test Results

## SUITE_WRITER: LibreOffice Writer

* **TEST_WRITER_KM (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the khmer_angkor keyboard. 2. Switched to khmer_angkor keyboard. 3. Opened the LibreOffice Writer document. 4. Typed "xEjmr>" letters. 5. Noticed that it showed the correct output.


* **TEST_WRITER_KM_BS (PASSED):** Retested with the attached PR build (Keyman 17.0.245-alpha-local (package version 17.0.245-1~PR-10480-2146.1+jammy1) on Ubuntu 22.04 Jammy Jelly fish Linux OS (vm) (Ubuntu on Xorg - x11) and here is my observation: 1. Installed the khmer_angkor keyboard. 2. Switched to khmer_angkor keyboard. 3. Opened the LibreOffice Writer document. 4. Typed "xEjmr" letters. 5. It showed the correct output.  6. With the output "ខ្មែរ" still being displayed, pressed the Backspace key 5 times. Verified that after each keystroke a part of the string is removed and after the 5th backspace the entire word was deleted.

**Comment by ermshiperete** [comment]
SUITE_TERMINAL: gnome-terminal

- TEST_TERMINAL_IPA (SKIPPED): Fails because of #10536

**Pull request merged**